src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -3090,15 +3117,30 @@ NakedWin32Handle templateFileWin32Handle
             );
 
             [DllImport(PinvokeDllNames.WriteConsoleDllName, SetLastError = true, CharSet = CharSet.Unicode)]
-            internal static extern bool WriteConsole
+            private static extern unsafe bool WriteConsole
             (
                 NakedWin32Handle consoleOutput,
-                string buffer,
+                char* buffer,
                 DWORD numberOfCharsToWrite,
                 out DWORD numberOfCharsWritten,
                 IntPtr reserved
             );
 
+            internal static unsafe bool WriteConsole
+            (
+                NakedWin32Handle consoleOutput,
+                ReadOnlySpan<char> buffer,
+                DWORD numberOfCharsToWrite,
+                out DWORD numberOfCharsWritten,
+                IntPtr reserved
+            )
+            {
+                fixed (char* bufferPtr = &MemoryMarshal.GetReference(buffer))
+                {
+                    return WriteConsole(consoleOutput, bufferPtr, numberOfCharsToWrite, out numberOfCharsWritten, reserved);
+                }
```

Comment:
This obviously complexes the code. Actually, all changes here make code more complex. Do we have measurable perf improvement to justify the changes? I don't see a large amount of string manipulation here (please point it out if I miss them). Note that many string APIs and `StringBuilder` already received perf improvements in 2.1.

Reply:
It looks like complement - I'm not sure I can create a _complex_ code :smile:

Exactly the pattern I found in CoreFX. This is due to C# language limitations.

Really the code isn't getting any more complicated. Rather on the contrary it becomes more complete.
By analyzing the code to clean up psl I found that in this place we make a lot of kernel calls and extra locks because of the multitude of:
```c#
WriteToConsole(text, true);
WriteToConsole(Crlf, true);
```
I was very surprised that we don't have a WriteLine() method to write to a host and use many allocations like `value + Crlf` - so I just implemented the method. 
The only complicity of the code is that I had to modify three levels of interfaces. So really the PR is very simple.

> Note that many string APIs and StringBuilder already received perf improvements in 2.1.

Yes, I saw up to 1100000 allocations in VS Profiler at pwsh.exe 6.0 startup and up to 850000 allocation at 6.1 (already on .Net 2.1) startup. Nevertheless, I still see a lot of allocations. The profile show that up to 36% allocations is chars and strings. 
This shows that CoreCLR does not solve all allocation problems and we should continue to consume resources cautiously and to optimize our code.

Reply:
I'm not convinced whether the startup is the right scenario to measure for changes in this PR, but still, for the startup scenario, what's the improvement in object allocation with your changes?

Reply:
No, it was only sample that 2.1 reduce allocations but not remove its at all. I don't know how measure wins in the PR - VS Profiler is crashed after first run. I look https://benchmarkdotnet.org/ but it takes a lot of effort. At least in the first steps.

Reply:
I attached a file with simple perf tests. I saw win ~2 second on 100000 outputs.
[measure_conhost.txt](https://github.com/PowerShell/PowerShell/files/2080705/measure_conhost.txt)

Reply:
Do we really need any of this code anymore? It was written back in V1 when the `Console` API was very limited. I suspect that we could remove most of the code with just a thin wrapper around the .NET APIs. (Which, in fact, were somewhat inspired by this code.)

Reply:
I agree that this code would have to be cleaned more thoroughly. The problem is that we do not have tests for the console code. Therefore, I would prefer to make small changes step by step so as not to break something. I intend to continue working in this direction (if only you do not chop it in one your PR :smile:).